src/lstm_layer.cpp

```diff
@@ -1515,3 +1515,505 @@ void LSTM::preinit_layer() {
+// SLSTM: LSTM layer with smoother
+////////////////////////////////////////////////////////////////////////////////
+
+std::string SLSTM::get_layer_info() const
+/*
+ */
+{
+    return "SLSTM(" + std::to_string(this->input_size) + "," +
+           std::to_string(this->output_size) + ")";
+}
+
+std::string SLSTM::get_layer_name() const
+/*
+ */
+{
+    return "SLSTM";
+}
+
+LayerType SLSTM::get_layer_type() const
+/*
+ */
+{
+    return LayerType::SLSTM;
+}
+
+void lstm_cov_cell_states_smoother(std::vector<float> &var_c_prev,
+                                   std::vector<float> &mu_f_ga,
+                                   std::vector<std::vector<float>> &cov_cc)
+/*
+ */
+{
+    std::vector<float> Cc_c(var_c_prev.size());
+    for (int i = 0; i < var_c_prev.size(); i++) {
+        // cov(c_{t-1},c_{t})
+        Cc_c[i] = var_c_prev[i] * mu_f_ga[i];
+    }
+    cov_cc.push_back(Cc_c);
+}
+
+void lstm_cov_hidden_cell_states_smoother(
+    std::vector<float> &var_c_prior, std::vector<float> &mu_o_ga,
+    std::vector<float> &jcb_ca, std::vector<std::vector<float>> &cov_hc)
+/*
+ */
+{
+    std::vector<float> Ch_c(var_c_prior.size());
+    for (int i = 0; i < var_c_prior.size(); i++) {
+        // cov(h_{t},c_{t})
+        Ch_c[i] = var_c_prior[i] * jcb_ca[i] * mu_o_ga[i];
+    }
+    cov_hc.push_back(Ch_c);
+}
+
+void lstm_cov_hidden_states_smoother(
+    std::vector<float> &mw, std::vector<float> &Jf_ga,
+    std::vector<float> &mi_ga, std::vector<float> &Ji_ga,
+    std::vector<float> &mc_ga, std::vector<float> &Jc_ga,
+    std::vector<float> &mo_ga, std::vector<float> &Jo_ga,
+    std::vector<float> &mu_h_prior, std::vector<float> &mu_h_prev,
+    std::vector<float> &var_h_prev, std::vector<float> &mc_prev,
+    std::vector<float> &mca, std::vector<float> &Jca, int w_pos_f, int w_pos_i,
+    int w_pos_c, int w_pos_o, int no, int ni, int start_idx, int end_idx,
+    std::vector<std::vector<float>> &cov_hh)
+/*
+ */
+{
+    std::vector<float> Ch_h(end_idx * end_idx, 0);
+    float Czz_f, Czz_i, Czz_c, Czz_o;
+    int m;
+
+    for (int t = start_idx; t < end_idx; t++) {
+        for (int j = 0; j < no; j++) {
+            // Forget gate
+            Czz_f = var_h_prev[t] * Jca[j] * mo_ga[j] * Jf_ga[j] *
+                    mw[(ni + no) * j + t + ni + w_pos_f] * mc_prev[j];
+
+            // Input gate
+            Czz_i = var_h_prev[t] * Jca[j] * mo_ga[j] * Ji_ga[j] *
+                    mw[(ni + no) * j + t + ni + w_pos_i] * mc_ga[j];
+
+            // Cell state gate
+            Czz_c = var_h_prev[t] * Jca[j] * mo_ga[j] * Jc_ga[j] *
+                    mw[(ni + no) * j + t + ni + w_pos_c] * mi_ga[j];
+
+            // Output gate
+            Czz_o = var_h_prev[t] * Jo_ga[j] *
+                    mw[(ni + no) * j + t + ni + w_pos_o] * mca[j];
+
+            // Updating quantities
+            m = t * no + j;
+            Ch_h[m] = Czz_f + Czz_i + Czz_c + Czz_o;
+        }
+    }
+    cov_hh.push_back(Ch_h);
```

Comment:
Could we move this variable in the loop to replace `Ch_h`? allocating a vector on-fly is not a desired way in this case because this function has no `return`